### PR TITLE
Added consistency improvement in list, details and create page of node credentials.

### PR DESCRIPTION
**Pull request by samshinde**
Signed-off-by: samshinde <ashinde@chef.io>

### :nut_and_bolt: Description: What code changed, and why?
consistency improvements for the list, details and create of node credential.
1. create node credential using the modal popup
2. Delete node credential modal pop-up with notification
3. Details tab which has update and reset buttons of each type(ssh, winrm, sudo).
4. ssh has a dropdown for password or rsa key, since only 1 is required not both

### :chains: Related Resources
#3716 
### :+1: Definition of Done
I have added consistency changes for the `create  Node Credential `, `list`, `update` and `delete  Node Credential ` with UI changes.
### :athletic_shoe: How to Build and Test the Change
visit the Settings >> Node Credential page,
create Node Credential  -  Create node credential according to the type you need.
delete Node Credential  - click to any Node Credential left side three-dot menu `...`  where you can see the delete Node Credential link to test delete the Node Credential.
```
STEP 1
inside the hab studio

[default:/src:0]# build components/automate-ui-devproxy/
[default:/src:0]# start_automate_ui_background
[default:/src:0]# start_all_services

STEP 2
open new window
go to automate UI path

$ cd components/automate-ui
and run the command 
```
### :white_check_mark: Checklist

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [ ] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
create
create-success
create-test
create-winrm
delete
delete-success
details
list-page
reset-tab
ssh-details
ssh-reset-tab
sudo-create
sudo-details
sudo-resettab

### CODE REVIEW

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+              <div class="version-dropdown">
+                <chef-select #ty [value]="passwordSelected" (change)="selectChangeHandlers(ty.value)" >
+                  <chef-option *ngFor="let list of types" [value]="list">{{ list }}</chef-option>
+                </chef-select>
+              </div>
+            </div>
+            <div class="name-margin">
+              <chef-form-field *ngIf="passwordSelected === 'Password'">
+                <label for="sshPassword">SSH password <span class="t-critical" aria-hidden="true">*</span></label>
+                <input chefInput id="sshPassword" formControlName="password" type="password" placeholder="password for private ssh" (keyup)="handleInput($event)" autocomplete="off"/>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:65]:
```suggestion
                <input chefInput id="sshPassword" formControlName="password" type="password"  (keyup)="handleInput($event)" autocomplete="off"/>
```

File: components/automate-ui/src/app/entities/node-credentials/node-credential.effects.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      map((resp: NodeCredentialResponse) => new CreateNodeCredentialSuccess(resp)),
+      catchError((error: HttpErrorResponse) =>
+      observableOf(new CreateNodeCredentialFailure(error))))));
+
+  @Effect()
+  createNodeCredentialSuccess$ = this.actions$.pipe(
+    ofType(NodeCredentialActionTypes.CREATE_SUCCESS),
+    map(({ payload: { } }: CreateNodeCredentialSuccess) => new CreateNotification({
+      type: Type.info,
+      message: 'Created node credential '
```
Comment (susanev) [inline components/automate-ui/src/app/entities/node-credentials/node-credential.effects.ts:94]:
```suggestion
      message: 'Created node credential.'
```

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 3]
```diff
@@ -0,0 +1,147 @@
+<chef-modal [visible]="visible" (closeModal)="closeCreateModal()">
+  <h2 slot="title">Create credential</h2>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:2]:
```suggestion
  <h2 slot="title">Create Credential</h2>
```

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        </ng-container>
+
+        <div id="button-bar">
+          <chef-button primary
+            id="create-button-object-modal"
+            data-cy="save-button"
+            [disabled]="!createNodeCredForm?.dirty"
+            (click)="createUser()">
+            <chef-loading-spinner *ngIf="creatingUser"></chef-loading-spinner>
+            <span *ngIf="!creatingUser">Create credential</span>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:140]:
```suggestion
            <span *ngIf="!creatingUser">Create Credential</span>
```

**Comment by susanev** [comment]
@samshinde both the details and reset credential forms have issues with the spacing between the inputs, please use the same amount of space as the other forms, i believe its 30px

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            <label>
+              <span class="label"> Credential Type <span aria-hidden="true">*</span></span>
+              <span class="custom-dropdown">
+                <select id="expression-type-dropdown"
+                        #cred
+                        (change)="updateFormDisplay(cred.value)"
+                        formControlName="type"
+                        [(value)]="selected">
+                  <option value='ssh'>SSH</option>
+                  <option value='winrm'>WinRm</option>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:35]:
```suggestion
                  <option value='winrm'>WinRM</option>
```

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+              </chef-form-field>
+            </div>
+          </form>
+        </ng-container>
+        <ng-container *ngIf="selected === 'winrm'">
+          <form [formGroup]="winrmForms">
+            <div class="name-margin">
+              <chef-form-field>
+                <label>
+                  <span class="label">Winrm Username <span aria-hidden="true">*</span></span></label>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:89]:
```suggestion
                  <span class="label">WinRM Username <span aria-hidden="true">*</span></span></label>
```

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                  <input chefInput id="winrmUsername-input" formControlName="username" type="text" (keyup)="handleInput($event)" autocomplete="off"/>
+                  <chef-error *ngIf="(winrmForms.get('username').hasError('required') || winrmForms.get('username').hasError('pattern')) && winrmForms.get('username').dirty">
+                    Username Key is required
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:92]:
```suggestion
                    WinRM username is required.
```

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+                  </chef-error>
+                </chef-form-field>
+            </div>
+            <div class="name-margin">
+                <chef-form-field >
+                  <label>
+                    <span class="label">Winrm password <span aria-hidden="true">*</span></span></label>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:99]:
```suggestion
                    <span class="label">WinRM Password <span aria-hidden="true">*</span></span></label>
```

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                  <input chefInput id="winrmPassword-input" formControlName="password" type="password" placeholder="password for private ssh" (keyup)="handleInput($event)" autocomplete="off"/>
+                  <chef-error *ngIf="(winrmForms.get('password').hasError('required') || winrmForms.get('password').hasError('pattern')) && winrmForms.get('password').dirty">
+                    Winrm Password is required
```
Comment (susanev) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html:102]:
```suggestion
                    WinRM Password is required
```

**Comment by samshinde** [comment]
> @samshinde both the details and reset credential forms have issues with the spacing between the inputs, please use the same amount of space as the other forms, I believe its 30px

@susanev updates the space between the inputs for the details and reset the credential tab.

**Review by msorens — COMMENTED** [review]
Some initial feedback; more later...

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+
+  public getLastModified(lastModifiedDate): string {
+    return moment.utc(lastModifiedDate).format(DateTime.RFC2822);
+  }
+
+  public inUseMessage(): string {
+    return (this.nodeCredentialToDelete && this.nodeCredentialToDelete.data.length > 0) ?
+      'and will disrupt access for all attached members' : '';
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts:142]: This wording (copied from PolicyListComponent) is not appropriate here; policies have members but node credentials do not. What are the data associated with a node credential?

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials.module.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { RouterModule } from '@angular/router';
+import {
+  FormsModule,
+  ReactiveFormsModule
+} from '@angular/forms';
+import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
+
+import { ChefComponentsModule } from '../../../components/chef-components.module';
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials.module.ts:10]: Should not use parent (..) notation in imports:
```suggestion
import { ChefComponentsModule } from 'app/components/chef-components.module';
```
Similar changes for other imports.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.html [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        (deleteClicked)="deleteNodeCredential()"
+        objectAction="Delete">
+      </app-delete-object-modal>
+      <section class="page-body">
+        <chef-loading-spinner *ngIf="nodesListLoading" size="50"></chef-loading-spinner>
+        <ng-container *ngIf="!nodesListLoading">
+          <chef-toolbar>
+            <chef-button primary (click)="openCreateModal()" data-cy="team-create-button">Create Credential
+            </chef-button>
+          </chef-toolbar>
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.html:28]: Would like to wrap in an authorization guard. Are you familiar with this notion? If not, search through the code base for other examples to see how they are used.
```suggestion
        <app-authorized [allOf]="['/api/v0/secrets', 'post']">
          <chef-toolbar>
            <chef-button primary (click)="openCreateModal()" data-cy="team-create-button">Create Credential
            </chef-button>
          </chef-toolbar>
        </app-authorized>
```

**Review by msorens — COMMENTED** [review]
A lot of good work here!
Note that I have not exercised the code; my notes are from code-reading only.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                  </chef-error>
+                </chef-form-field>
+            </div>
+            <div class="name-margin">
+                <chef-form-field >
+                  <label>
+                    <span class="label">Command Line options </span></label>
+                  <input chefInput id="command-input" formControlName="options" 
+                  type="command" 
+                  placeholder="same as the sudo command line options you use in the terminal, e.g: -n -u ubuntu"  (keyup)="handleInput($event)" autocomplete="off"/>
```
[the remark this answers is not in the document]
Comment (samshinde) [reply components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.html]: Yes, the text will not appear if the user has a smaller window so I have removed that placeholder for know. and will look for the solution.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+    this.store.dispatch(new CreateNodeCredential(userCreateReq));
+  }
+
+  handleUsernameInput(event: KeyboardEvent): void {
+    if (this.isNavigationKey(event)) {
+      return;
+    }
+    this.conflictError = false;
+  }
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/create-node-credential-modal/create-node-credential-modal.component.ts:127]: Delete this duplicate function: you already have it as `handleInput` below.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    this.tabValue = event.target.value;
+    this.router.navigate([this.url.split('#')[0]], { fragment: event.target.value });
+  }
+
+  selectChangeHandlers(id: string): void {
+    this.passwordSelected = id;
+  }
+
+  saveNodeCredential(data: NodeObject): void {
+    let credBody;
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts:124]: Can we add a type for this?

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+      this.resetSuccessful = false;
+      this.resetInProgress = true;
+      credBody = this.saveCred.getNodeCredentialCreate(data);
+    }
+
+    this.store.dispatch(new UpdateNodeCredential(credBody));
+    this.resetForm.reset();
+  }
+
+  getNodeCred() {
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts:145]: 1. add return type
2. Would suggest renaming to, e.g. "setupNodeSubscriptions" to make it clear that it is an initialization function

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+        if (this.tabValue === 'details') {
+          this.saveInProgress = false;
+          this.saveSuccessful = (state === EntityStatus.loadingSuccess);
+          this.updateForm.markAsPristine();
+        } else {
+          this.resetInProgress = false;
+          this.resetSuccessful = (state === EntityStatus.loadingSuccess);
+          this.resetForm.markAsPristine();
+        }
+        this.getNodeCred();
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts:109]: Delete this--method is already called on L84 and should only be called once.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+  public orderBy: NodeCredentialOrder;
+  public nodeCredentialToDelete: NodeCredential;
+  public createModalVisible = false;
+  public creatingNodeCredential = false;
+  public deleteModalVisible = false;
+  public nodesListLoading = true;
+  public messageModalVisible = false;
+  public nodesListSort = 'last_contact';
+  public nodesListSortOrder = 'DESC';
+  public nodesListFilters = [{ key: 'type', values: ['ssh', 'winrm', 'sudo'] }];
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts:46]: Unused; delete.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    this.layoutFacade.showSidebar(Sidebar.Settings);
+    this.getNodeList();
+
+  }
+
+  openCreateModal(): void {
+    this.openUserModal.emit(true);
+  }
+
+  orderFor(column) {
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts:67]: Add argument type and return type

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+    return column === this.sortBy ? this.orderBy : 'none';
+  }
+
+  handleSortToggle({ detail: sortParams }) {
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts:71]: Add return type

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+  startNodeCredentialDelete($event: MatOptionSelectionChange, p: NodeCredential): void {
+    if ($event.isUserInput) {
+      this.nodeCredentialToDelete = p;
+      this.deleteModalVisible = true;
+    }
+  }
+
+  getNodeList() {
+    this.instanceNodeCredentials$ = this.store.pipe(select(allCredentials));
+    this.store.select(nodeCredentialListState).pipe(
```
Comment (msorens) [inline components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credentials-list/node-credential-list.component.ts:99]: This subscription needs to be only invoked once; you are invoking it multiple times through multiple calls of `getNodeList`. Need to separate out that setup code from other code in this function (like the dispatch).

**Review by msorens — COMMENTED** [review]
Thanks for the updates; did another review pass with some additional notes.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts (hunk 17 above)
Comment (samshinde) [reply components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts]: @msorens Actually we don't have any response after calling the patch Api so for updating the node credential we need to call it as the same happening in creating success.

**Comment by samshinde** [comment]
@msorens regarding the Conflict error test case for now I have disabled the test case and looking for a solution for the same.

File: components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts (hunk 17 above)
Comment (msorens) [reply components/automate-ui/src/app/pages/+compliance/+node-credentials/node-credential-details/node-credential-details.component.ts]: No, this violates subscription lifecycle management. The reason I asked you even to rename it to `setupNodeSubscriptions` is that it is **not** performing any actions--it is simply creating subscriptions and we do not want to setup the same subscription call twice--that simply means the subscription handler will run twice **every** time the data is emitted.

**Comment by msorens** [comment]
Re-reviewed once more; just the one item that needs addressing (https://github.com/chef/automate/pull/4323#discussion_r496309743), then looks good to go!

**Review by susanev — APPROVED** [review]

**Comment by samshinde** [comment]
> Re-reviewed once more; just the one item that needs addressing ([#4323 (comment)](https://github.com/chef/automate/pull/4323#discussion_r496309743)), then looks good to go!

@msorens Yes, I have updated the code and resolve the two API call issue.

**Review by msorens — APPROVED** [review]

**Pull request merged**